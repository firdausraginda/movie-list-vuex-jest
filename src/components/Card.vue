<template>
  <div class="card col-4">
    <div class="card-body">
      <img
        class="card-img-top"
        :src="dataCard.imgFilm"
        alt="Card image cap"
      >
      <h5 class="card-title">{{cutTextJudul(dataCard.judulFilm)}}</h5>
      <p>{{cutTextDeskripsi(dataCard.deskripsiFilm)}}</p>
      <a
        class="btn btn-primary moveToDetail"
        @click="moveToDetail()"
      >Detail</a>
      <a
        v-if="dataCard.choosen == false"
        class="btn chooseDataFilm"
        @click="chooseDataFilm()"
      >Pilih</a>
      <a
        v-if="dataCard.choosen == true && this.$route.name == 'personal'"
        class="btn deleteDataFilm"
        @click="deleteDataFilm()"
      >Hapus</a>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {};
  },
  props: {
    dataCard: {
      type: Object,
      default: null
    }
  },
  methods: {
    cutTextJudul(text) {
      return `${text.substr(0, 30)}`;
    },
    cutTextDeskripsi(text) {
      return `${text.substr(0, 100)}..`;
    },
    moveToDetail() {
      this.$router.push({ path: `/detailFilm/${this.dataCard.idFilm}` });
    },
    chooseDataFilm() {
      this.$store.dispatch("addDataFilmChoosen", this.dataCard);
    },
    deleteDataFilm() {
      this.$store.dispatch("deleteDataFilmChoosen", this.dataCard);
    }
  }
};
</script>

<style lang="scss" scoped>
.btn-primary {
  color: white !important;
}
.card {
  border: none;
  text-align: left;
  .card-img-top {
    border-radius: 5px;
    height: 200px;
    object-fit: cover;
    margin-bottom: 20px;
  }
  .card-title {
    font-weight: bold;
  }
}
</style>