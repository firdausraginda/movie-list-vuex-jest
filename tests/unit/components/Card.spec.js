import { shallowMount, createLocalVue } from '@vue/test-utils'
import Card from '../../../src/components/Card.vue'
import Vuex from 'vuex'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('Card.spec.js', () => {
    let wrapper;
    let actions;
    let store;
    beforeEach(() => {
        actions = {
            deleteDataFilmChoosen: jest.fn(),
            addDataFilmChoosen: jest.fn()
        }
        store = new Vuex.Store({
            actions
        })
        wrapper = shallowMount(Card, {
            store,
            localVue,
            propsData: {
                dataCard: { idFilm: 'gans', choosen: false, imgFilm: 'https://upload.wikimedia.org/wikipedia/en/thumb/5/53/Black_mirror_bandersnatch_poster.jpg/220px-Black_mirror_bandersnatch_poster.jpg', judulFilm: 'Black Mirror: Bandersnatch', deskripsiFilm: 'Black Mirror: Bandersnatch adalah film interaktif tahun 2018 dalam seri antologi fiksi ilmiah Black Mirror. Film ini ditulis oleh pencipta serinya, Charlie Brooker, dan disutradarai oleh David Slade. Netflix merilisnya sebagai film terpisah pada tanggal 28 Desember 2018.' }
            },
            methods: {
                cutTextJudul(text) {
                    return `${text.substr(0, 30)}`;
                },
                cutTextDeskripsi(text) {
                    return `${text.substr(0, 100)}..`;
                },
                moveToDetail() {
                    this.$router.push({ path: `/detailFilm/${this.dataCard.idFilm}` });
                },
                chooseDataFilm() {
                    this.$store.dispatch("addDataFilmChoosen", this.dataCard);
                },
                deleteDataFilm() {
                    this.$store.dispatch("deleteDataFilmChoosen", this.dataCard);
                }
            }
        })
    })

    test('is Card component', () => {
        expect(wrapper.is(Card)).toBe(true)
        expect(wrapper.is('.card')).toBe(true)
    })

    test('Card property has object type of data with value ', () => {
        expect(wrapper.vm.dataCard).not.toBeNull()
        expect(wrapper.vm.dataCard.idFilm).toBe('gans')
        expect(typeof wrapper.vm.dataCard).toBe('object')
        expect(wrapper.vm.dataCard.imgFilm).toEqual('https://upload.wikimedia.org/wikipedia/en/thumb/5/53/Black_mirror_bandersnatch_poster.jpg/220px-Black_mirror_bandersnatch_poster.jpg')
        expect(wrapper.vm.dataCard.idFilm).toEqual('gans')
        expect(wrapper.vm.dataCard.judulFilm).toEqual('Black Mirror: Bandersnatch')
    })

    test('has the expected html structure', () => {
        expect(wrapper.element).toMatchSnapshot()
    })

    test('check the length of judul film after "cutTextJudul" being called', () => {
        let getJudulFilm = wrapper.vm.dataCard.judulFilm
        let cutJudulFilm = wrapper.vm.cutTextJudul(getJudulFilm)
        expect(cutJudulFilm.length).toBeLessThanOrEqual(32)
    })

    test('check the length of deskripsi film after "cutTextDeskripsi" being called', () => {
        let getDesFilm = wrapper.vm.dataCard.deskripsiFilm;
        let cutDesFilm = wrapper.vm.cutTextDeskripsi(getDesFilm)
        expect(cutDesFilm).toHaveLength(102)
    })

    test('dispatches "addDataFilmChoosen" when "chooseDataFilm" being called', () => {
        let btnChooseDataFilm = wrapper.find('.chooseDataFilm')
        btnChooseDataFilm.trigger('click')
        expect(actions.addDataFilmChoosen).toHaveBeenCalled()
    })

    test('dispatches "deleteDataFilmChoosen" when "deleteDataFilm" being called', () => {
        let btnDeleteDataFilm = wrapper.find('.deleteDataFilm')
        expect(btnDeleteDataFilm.exists()).toBe(true)
        // btnDeleteDataFilm.trigger('click')
        // expect(actions.deleteDataFilmChoosen).toHaveBeenCalled()
    })

})