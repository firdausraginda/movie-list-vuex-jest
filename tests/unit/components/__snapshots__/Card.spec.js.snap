// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card.spec.js has the expected html structure 1`] = `
<div
  class="card col-4"
>
  <div
    class="card-body"
  >
    <img
      alt="Card image cap"
      class="card-img-top"
      src="https://upload.wikimedia.org/wikipedia/en/thumb/5/53/Black_mirror_bandersnatch_poster.jpg/220px-Black_mirror_bandersnatch_poster.jpg"
    />
     
    <h5
      class="card-title"
    >
      Black Mirror: Bandersnatch
    </h5>
     
    <p>
      Black Mirror: Bandersnatch adalah film interaktif tahun 2018 dalam seri antologi fiksi ilmiah Black ..
    </p>
     
    <a
      class="btn btn-primary moveToDetail"
    >
      Detail
    </a>
     
    <a
      class="btn chooseDataFilm"
    >
      Pilih
    </a>
     
    <!---->
  </div>
</div>
`;
